File: Recursion/90M. Subsets II.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        temp = []
        check = [0]*len(nums)
        nums.sort()
        
        def isValid(nums, check, i):
            if check[i]==1:
                return False
            if i>0 and nums[i]==nums[i-1] and check[i-1]==0:
                return False
            return True
        
        def backtrack(nums, res, temp, check, start):
            res.append(list(temp))
            for i in range(start, len(nums)):
                if isValid(nums, check, i):
                    check[i]=1
                    temp.append(nums[i])
                    backtrack(nums, res, temp, check, i+1)
                    temp.pop()
                    check[i]=0
        backtrack(nums, res, temp, check, 0)
        return res
    
    def subsetsWithDup(self, nums):
        self.ans = [[]]
        tmp = []
        check = [0]*len(nums)
        self.helper(sorted(nums), 0, check, tmp)
        return self.ans
    
    def valid(self, nums, check, i):
        if check[i]==1:
            return False
        elif i>0 and nums[i]==nums[i-1] and check[i-1]==0:
            return False
        return True
    
    def helper(self, nums, start, check, tmp):
        if start==len(nums):
            return
        for i in range(start, len(nums)):
            if self.valid(nums, check, i):
                tmp.append(nums[i])
                self.ans.append(copy.deepcopy(tmp))
                check[i]=1
                self.helper(nums, i+1, check, tmp)
                check[i]=0
                tmp.pop()

File: Recursion/90M. Subsets II.cpp
/*
90M. Subsets II
Array, Backtracking

Description:
Given an integer array nums that may contain duplicates, return all possible subsets (the power set).
The solution set must not contain duplicate subsets. Return the solution in any order.

Example 1:
Input: nums = [1,2,2]
Output: [[],[1],[1,2],[1,2,2],[2],[2,2]]

Example 2:
Input: nums = [0]
Output: [[],[0]]

Constraints:
1 <= nums.length <= 10
-10 <= nums[i] <= 10

Similar Questions:
Subsets - Mediuim
*/


/*
Solution: Similar to 47M. Permutations II
          Sort the array first so as to put duplicates together.
          Since there are duplicated values in the array, we need a vector "check" to check if the value has been added. 
*/
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> temp;
        vector<int> check(nums.size(), 0);
        sort(nums.begin(), nums.end());
        subsets(nums, temp, check, 0);
        return res;
    }
    void subsets(vector<int>& nums, vector<int>& temp, vector<int>& check, int start){
        res.push_back(temp);
        for(int i=start; i<nums.size(); ++i){
            if(isValid(nums, check, i)){
                temp.push_back(nums[i]);
                check[i] = 1;
                subsets(nums, temp, check, i+1);
                check[i] = 0;
                temp.pop_back();
            }
        }
    }
    bool isValid(vector<int>& nums, vector<int>& check, int& i){
        if(check[i]==1){
            return false;
        }
        if(i>0 && nums[i]==nums[i-1] && check[i-1]==0){
            return false;
        }
        return true;
    }
};
